In [121]:
import os

# Computational Theory Tasks

## Task 1: Binary Representations

This task involves implementing four functions aimed to manipulate binary data, this is fundemental to many crythographic algorithms. Bitwise Operations allow for the manipulation of individual bits of data, these operations are applied directly to binary representations of numbers.  In cryptography bitwise operations are used to ensure confusion and diffusion in data.  Some common uses of these operations are in Hash functions, Encryption algorithms, Blockchain aswell as many others.

---------------------------------------------------------------------------

### 1. 'rotl(x, n=1)' 
This function rotates bits of a 32-bit unsigned integer 'x' to the left by 'n' places this is effectively shifting bits and wrapping the leftmost bits back into the rightmost position. The most common use for these functions are in crypographic hash functions such as SHA-1 and SHA-256 to provide bitwise diffusion making it difficult to reverse hash.  They are also used in applications like cylic buffers or rotating registers in hardware.

In [122]:
def rotl(x, n=1):
    x &= 0xFFFFFFFF  #ensure x is a 32-bit unsigned integer
    return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF #rotate left by n bits

#### Step 1 Ensure x is a 32-bit unsigned integer.
- `0xFFFFFFFF` is the hexidecimal constant that represents a 32-bit mask.

- The bitwise AND (&) operation ensures that x only has the lower 32 bits, preventing unexpected large numbers.
#### Step 2 Rotate left by n bits
- `(x << n)` : Left shift moves bits n places to the left, filling with 0s, bits moved beyond 32nd bit are lost

- `(x >> (32 - n))` : Right shift moves the lost bits back to rightmost positions 

- `|` : Bitwise OR combines the shifted values.
### Step 3 Ensure Result is 32-bit
- `& 0xFFFFFFFF` : This ensures result stays within 32-bit unsigned integer range

In [123]:
#example usage:
x = 0b01
n = 3
result = rotl(x, n)
print(f"Original: {x:032b}")
print(f"Rotated Left {n} positions: {result:032b}")

Original: 00000000000000000000000000000001
Rotated Left 3 positions: 00000000000000000000000000001000


---------------------------------------------------------------------------

### 2. 'rotr(x n=1)'
This function rotates bits of a 32-bit unsigned integer 'x' to the right 'n' places this is effectively shifting bits and wrapping the rightmost bits back into the leftmost position. This is similar to the 'rotl' function with the same use cases.

In [124]:
def rotr(x, n=1):
    x &= 0xFFFFFFFF #ensure x is a 32-bit unsigned integer
    return ((x >> n) | (x << (32 - n))) & 0xFFFFFFFF #rotate right by n bits

#### Step 1 Ensure x is a 32-bit unsigned integer.
- `0xFFFFFFFF` is the hexidecimal constant that represents a 32-bit mask.

- The bitwise AND (&) operation ensures that x only has the lower 32 bits, preventing unexpected large numbers.
#### Step 2 Rotate right by n bits
- `(x >> n)` : Right shift moves bits n places to the right, filling with 0s, bits moved beyond 32nd bit are lost

- `(x << (32 - n))` : Left shift moves the lost bits back to leftmost positions 

- `|` : Bitwise OR combines the shifted values.
### Step 3 Ensure result is 32-bit
- `& 0xFFFFFFFF` : This ensures result stays within 32-bit unsigned integer range

In [125]:
#example usage:
x = 0b01
n = 3
result = rotr(x, n)
print(f"Original: {x:032b}")
print(f"Rotated Right by {n} positions: {result:032b}")

Original: 00000000000000000000000000000001
Rotated Right by 3 positions: 00100000000000000000000000000000


---------------------------------------------------------------------------

### 3. 'def ch(x, y, z)'
This function chooses bits from two values: y and z, based off the corresponding bits in x. This works by choosing the corresponding bit in y if the x bit is 1; and choosing the bit corresponding bit in z if x is 0. This is a core compononent in the SHA-2(Secure Hash Algorithm) its used in the message schedule, it works with the bitwise operations to choose part of data from different sources based on values of x. Its also helps mix data from multiple sources to enhace unpredictability of cryptogrpahic transformations.

In [126]:
def ch(x, y, z):
     #choose y if x is true(1), otherwise choose z if x is false(0)
    return (x & y) ^ (~x & z)

#### Step 1 Select y where x is 1
- `(x & y)` : Bitwise AND(&) ensures that only the bits where x is 1 are taken from y
#### Step 2 Select z where x is 0
- `(~x & z) `: Bitwise NOT (~x) inverts bits in x, making the 1s to 0s and vice versa

- Bitwise AND(&) with z ensures only bits were x was originally 0 can be taken form z
#### Step 3 Combine results
- `(x & y) ^ (~x & z)` : Bitwise XOR(^) combines the two selections, ensuring bits from only one (y or z) is taken at each position

In [127]:
#example usage:
x = 0b1100  
y = 0b1010  
z = 0b0110 
result = ch(x, y, z)
print(f"x: {x:04b}, y: {y:04b}, z: {z:04b}")
print(f"ch(x, y, z): {result:04b}")

x: 1100, y: 1010, z: 0110
ch(x, y, z): 1010


---------------------------------------------------------------------------

### 4. 'maj(x, y, z)'
This function takes a majority vote of the bits in each position in x, y and z. Output has 1 in position i if at least two of x, y, z have 1s in position i otherwise the result is 0. This is commonly used in Cryptographic hash functions such as SHA-256 and ensures output is influenced by majority of bits, increasing diffusion, improving resistance to attacks. Its also used in certain Error correction Codes or correction schemes where majority of values are trusted, allowing erros to be flagged based on outliers.

In [128]:
def maj(x, y, z):
    #choose the majority of x, y, z
    return (x & y) ^ (x & z) ^ (y & z)

#### Step 1 Calculate the majority between x and y
- `(x & y)` : Bitwise AND(&) between x and y give the bits where both x and y have 1 in the same position

- This operation shows a majority of the two bits at each position
#### Step 2 Calculate the majority between x and z
- `(x & z)` : Bitwise AND(&) between x and z gives the bits where both x and z have 1 in the same position
#### Step 3 Calculate the majority between y and z
- `(y & z)` : Bitwise AND(&) between y and z gives the bits where both x and z have 1 in the same position
#### Step 4 Combine results with XOR(^)
- `(x & y) ^ (x & z) ^ (y & z)` : Bitwise XOR(^) combines the results of these three majoritys. The XOR operation ensures only the majority value of the three will be choosen

- Eg. If two of the three bits are 1, the result will be 1 fot that bit position


In [129]:
#example usage:
x = 0b1100  
y = 0b1010  
z = 0b0110 
result = maj(x, y, z)
print(f"x: {x:04b}, y: {y:04b}, z: {z:04b}")
print(f"maj(x, y, z): {result:04b}")

x: 1100, y: 1010, z: 0110
maj(x, y, z): 1110


---------------------------------------------------------------------------

### Test function
Testing each Binary operation function using hexdecimal number representations

In [130]:
#test cases
def test_binary_functions():
    #test rotl using hexadecimal numbers
    #0x01 = 0000 0001, 0x02 = 0000 0010
    assert rotl(0x01, 1) == 0x02
    #0x80000000 = 1000 0000 0000 0000 0000 0000 0000 0000, 0x00000001 = 0000 0000 0000 0000 0000 0000 0000 0001
    assert rotl(0x80000000, 1) == 0x00000001
    #0x12345678 = 0001 0010 0011 0100 0101 0110 0111 1000, 0x23456781 = 0010 0011 0100 0101 0110 0111 1000 0001 
    assert rotl(0x12345678, 4) == 0x23456781

    #test rotr using hexadecimal numbers
    #0x01 = 0000 0001, 0x02 = 0000 0010
    assert rotr(0x02, 1) == 0x01
    #0x80000000 = 1000 0000 0000 0000 0000 0000 0000 0000, 0x00000001 = 0000 0000 0000 0000 0000
    assert rotr(0x00000001, 1) == 0x80000000
    #0x12345678 = 0001 0010 0011 0100 0101 0110 0111 1000, 0x81234567 = 1000 0001 0010 0011 0100 0101 0110 0111
    assert rotr(0x12345678, 4) == 0x81234567

    #test ch using hexadecimal numbers(simple examples)
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0xFFFFFFFF, 0xAAAAAAAA, 0x55555555) == 0xAAAAAAAA #x is all ones, so the result is y
    #0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0x00000000, 0xAAAAAAAA, 0x55555555) == 0x55555555 #x is all zeros, so the result is z
    #0xF0F0F0F0 = 1111 0000 1111 0000 1111 0000 1111 0000, 0xAAAAAAAA = 1010 1010 1010 1010 1010 1010 1010 1010, 0x55555555 = 0101 0101 0101 0101 0101 0101 0101 0101
    assert ch(0xF0F0F0F0, 0xAAAAAAAA, 0x55555555) == 0xA5A5A5A5 #0xA5A5A5A5 = 1010 0101 1010 0101 1010 0101 1010 0101 (combination of y and z as x is mixed)

    #test maj using hexadecimal numbers (simple examples)
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000
    assert maj(0xFFFFFFFF, 0x00000000, 0x00000000) == 0x00000000 # majority is 0x00000000 so the result is 0x00000000
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0x00000000 = 0000 0000 0000 0000 0000 0000 0000 0000
    assert maj(0xFFFFFFFF, 0xFFFFFFFF, 0x00000000) == 0xFFFFFFFF # majority is 0xFFFFFFFF so the result is 0xFFFFFFFF
    #0xFFFFFFFF = 1111 1111 1111 1111 1111 1111 1111 1111, 0xF0F0F0F0 = 1111 0000 1111 0000 1111 0000 1111 0000
    assert maj(0xFFFFFFFF, 0xF0F0F0F0, 0xF0F0F0F0) == 0xF0F0F0F0 # majority is 0xF0F0F0F0 so the result is 0xF0F0F0F0

    print("All test cases passed successfully")

test_binary_functions()

All test cases passed successfully


---------------------------------------------------------------------------

### Task 1 References:
1. FIPS PUB 180-4: Secure Hash Standard (SHS) :https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf  
2. M. A. Hossain, M. M. Alam, and M. S. Alam, "Cryptography Encryption Technique Using Circular Bit Rotation in Binary Field," 2020 IEEE Region 10 Symposium (TENSYMP), Dhaka, Bangladesh, 2020: https://ieeexplore.ieee.org/document/9197845/
3. Rotate bits of a number: https://www.geeksforgeeks.org/rotate-bits-of-an-integer/
4. RealPython: Bitwise operators in python: https://realpython.com/python-bitwise-operators/  
5. Understanding Bitwise Operations in Python: Cryptography, Hashing, and Real-World Applications: https://www.linkedin.com/pulse/understanding-bitwise-operations-python-cryptography-hashing-woon-sm7ec/

---------------------------------------------------------------------------

## Task 2: Hash Functions

### Polynomial Rolling Hash Function
Hashing is common practice in the world of computer science used to map data, such as strings, to numerical values. A widely used approach for hasing strings is the polynomial rolling hash function.
This function is defined as:  
$$H(s) = (s_1 \times p^{(n-1)} + s_2 \times p^{(n-2)} + ... + s_n \times p^0) \mod m$$

- $H(s)$ is hash value of the string
- $s1, s2, sn$ are ASCII values of characters in the string
- $p$ is a prime base (commonly 31)
- $m$ is the modulus value (commonly 101)
- $n$ is the lenght of the string 

The Polynomial Rolling Hash Function has many use cases and is widely used in
- String Matching Algorithms
- Data Structures
- Plagiarism Detection
- Crypography

---------------------------------------------------------------------------

In [131]:
def hash(s):
    #initialise the hash function to 0
    hashval = 0

    #iterate over each character in the string
    for c in s:
        #compute hash using polynomial rolling hash function
        hashval = ord(c) + 31 * hashval
        print(f"Character: {c}, Hash: {hashval}")
    
    #apply modulo operation to keep hash within fixed range (0-100)
    return hashval % 101

---------------------------------------------------------------------------

### Code Explanation

- This variable will store the computed hash value, this will be happen as the function processes each charecter in the string.

In [132]:
hashval = 0

### Polynomial Rolling Hash Calculation
- This itereates over each charecter `c` in the string `s`. Each character in `s` contributes to the final hash.

- `ord(c)`: Converts the character `c` into ASCII value.

- `31 * hashval` : This takes the current hash and multiplies it by 31, this is a prime number commonly used for hash functions.

- `ord(c) + 31 * hashval` : The full line adds the ASCII of the character to a scaled hash value, this is to ensure each character incluences the hash in differently depending on the characters position.

In [133]:
#string to be hashed
s = "hello"

#interate over each character in the hello
for c in s:
    #compute hash using polynomial rolling hash function
    hashval = ord(c) + 31 * hashval

print(hashval)

99162322


### Modulo Operation (%101)
- This aims to ensure the final hash is in a specified range (0-100)

- This is to avoid faults like integer overflow, but also keeps the hash value compact.

- The prime modulous (101) is used as it helps distribute hash values evenly, which helps reduce collisions.

In [134]:
hashval % 101

17

---------------------------------------------------------------------------

### Example of how function works:

In [135]:
print(hash("hello"))

Character: h, Hash: 104
Character: e, Hash: 3325
Character: l, Hash: 103183
Character: l, Hash: 3198781
Character: o, Hash: 99162322
17


Iterate over each charecter in hello applying rolling hash calculation:
1. h (ASCII 104): hashval = 104
2. e (ASCII 101): hashval = 101 + 31 x 104 = 3325
3. l (ASCII 108): hashval = 108 + 31 x 3325 = 103183
4. l (ASCII 108): hashval = 108 + 31 x 103183 = 3198781
5. o (ASCII 111): hashval = 111 + 31 x 3198781 = 99162322

Apply modulo operation:
99162322 mod 101 = 17

"Hello" = 17

---------------------------------------------------------------------------

### Why use 31 as the Base?
1. Efficient Computation:
    - 31 is close to the powers of 2, this allows efficient multiplication and bitwise operations on many processors.
2. Empirical Evidence:
    - Many real-world applications use 31 due to it providing a good balance of speed and low collision probability.
3. Prime Number Property:
    - Prime numbers reduce chances of hash collisions occuring this is done by ensuring more uniform distribution of hash values

### Why use 101 as the Modulus?
1. Prime Numbers Improve Hash Distribution:
    - Prime modulus ensures values are distributed evenly this inturn reduces hash collisions
2. Keeps Hash Values Compact:
    - Modulo operations keep hash values in a range in this case that range is 0 -100 this makes it more efficient for hashing tables.

---------------------------------------------------------------------------

### Task 2 References:
1. String hasing using Polynomial rolling hash function: https://www.geeksforgeeks.org/string-hashing-using-polynomial-rolling-hash-function
2. String hashing: https://cp-algorithms.com/string/string-hashing.html
3. Why should hash functions use a prime number modulus?: https://www.designgurus.io/answers/detail/why-should-hash-functions-use-a-prime-number-modulus
4. Understanding Rolling Hash: A Key Component in String Matching Algorithms: https://medium.com/%40ggaappuu1234/understanding-rolling-hash-a-key-component-in-string-matching-algorithms-83236d8c4a20
5. Introduction to Rolling Hash – Data Structures and Algorithms: https://www.geeksforgeeks.org/introduction-to-rolling-hash-data-structures-and-algorithms/
6. On the mathematics behind rolling hashes and anti-hash tests: https://codeforces.com/blog/entry/60442

---------------------------------------------------------------------------

## Task 3: SHA256

In [ ]:
def sha256(file_path):
    #read the file contents
    with open(file_path, 'r') as f:
        binary_content = f.read().replace(' ', '')
    
    #convert binary string to bytes
    data = bytes(int(binary_content[i:i+8], 2) for i in range(0, len(binary_content), 8))
    
    #calculate original length in bits
    orig_len = len(data) * 8
    
    #start padding by adding the 1 bit (0x80 in hex)
    pad_data = data + b'\x80'
    
    #calculate padding length to make total length
    pad_len = (56 - (len(pad_data) % 64)) % 64
    
    #add zero padding
    pad_data += b'\x00' * pad_len
    
    #append original length as 64-bit big-endian unsigned integer
    pad_data += orig_len.to_bytes(8, 'big')
    
    #print padding in hex
    padding_hex = " ".join(f"{byte:02x}" for byte in pad_data[len(data):])
    print(padding_hex)
    

In [176]:
file_path = "test.txt"
sha256(file_path)

80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 18


b'abc\x80\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x18'

## Task 4: Prime Numbers

## Task 5: Roots

## Task 6: Proof of Work

## Task 7: Turing Machines

## Task 8: Computational Complexity

## End